kazu/steps/ner/llm_ner.py

```diff
@@ -0,0 +1,215 @@
+import json
+import logging
+from abc import abstractmethod, ABC
+from typing import Iterable, Optional, Any, cast
+
+import ahocorasick
+from kazu.data import Document, Entity, Section, MentionConfidence
+from kazu.steps import Step, document_iterating_step
+from kazu.utils.spacy_pipeline import BASIC_PIPELINE_NAME, SpacyPipelines, basic_spacy_pipeline
+from kazu.utils.utils import word_is_valid
+
+logger = logging.getLogger(__name__)
+
+LLM_RAW_RESPONSE = "llm_raw_response"
+
+
+class LLMModel(ABC):
+    def __call__(self, text: str) -> tuple[str, Optional[dict[str, str]]]:
+        """Call the LLM model with the given text and return the raw response and parsed
+        entities.
+
+        :param text: The text to pass to the LLM model.
+        :return: A tuple of the raw response and the found entities as a dict.
+        """
+        raw_response = self.call_llm(text)
+        parsed_response = self.parse_result(raw_response)
+        return raw_response, parsed_response
+
+    @abstractmethod
+    def call_llm(self, text: str) -> str:
+        """Call the LLM model with the given text and return the raw response.
+
+        :param text: The text to pass to the LLM model.
+        :return: The raw response from the LLM model.
+        """
+        pass
+
+    @staticmethod
+    def parse_result(result: str) -> Optional[dict[str, str]]:
+        """Parse the raw response from the LLM model into a dictionary of entities.
+
+        :param result: The raw response from the LLM model.
+        :return: A dictionary of entities and their class.
+        """
+        if "{}" in result:
+            return {}
+        try:
+            curly_braces_start = result.find("{")
+            square_start = result.find("[")
+            if square_start == -1 or square_start > curly_braces_start:
+                return cast(
+                    dict[str, str], json.loads(result[curly_braces_start : result.rfind("}") + 1])
+                )
+            else:
+                final = {}
+                for item in json.loads(result[square_start : result.rfind("]") + 1]):
+                    final.update(item)
+                return final
+        except Exception:
+            return None
+
+
+class AzureOpenAILLMModel(LLMModel):
```

Comment:
Thoughts on using something like LangChain instead? It basically already returns different objects depending on the platform that can be invoked in the same way to call the LLM. 

Also thinking whether the `parse_result` is necessary. Depending on what model you use - they should support structured output? You can create a pydantic class to define the output schema. In Openai it is https://platform.openai.com/docs/guides/structured-outputs/introduction or in Langchain  https://python.langchain.com/v0.1/docs/modules/model_io/chat/structured_output/

Reply:
Discussed in person. Langchain might be a bit overkill for now, but let's revisit if the requriements change

Reply:
also regarding structured outputs, this would be a cool feature but is not universally supported